Add support for excludepkgs option in tdnf.conf

sidcha opened the pull request:
This patch adds the config option "excludepkgs" from tdnf.conf and some
utils to make this possible neatly.

This also adds a new TDNFReAllocateMemory() that mimics realloc() but
also guarantees that the newly reallocated memory is also zero-ed to
meet the general tdnf codebase expectation.

Signed-off-by: Siddharth Chandrasekaran <csiddharth@vmware.com>

sidcha committed:
Add support for excludepkgs option in tdnf.conf

This patch adds the config option "exclude" from tdnf.conf and some
utils to make this possible neatly.

Signed-off-by: Siddharth Chandrasekaran <csiddharth@vmware.com>

sidcha commented:
CC: @keerthanakalyan, @sshedi, @jaankit, @tapakund, @oliverkurth

oliverkurth commented on client/defines.h:
```diff
@@ -79,6 +79,7 @@ typedef enum
 #define TDNF_CONF_KEY_PLUGIN_PATH         "pluginpath"
 #define TDNF_CONF_KEY_PLUGIN_CONF_PATH    "pluginconfpath"
 #define TDNF_PLUGIN_CONF_KEY_ENABLED      "enabled"
+#define TDNF_CONF_KEY_EXCLUDE             "exclude"
```
Should this be `excludepkgs`? I see `excludepkgs` but not `exclude` in https://dnf.readthedocs.io/en/latest/conf_ref.html .

sidcha replied:
This was passed down to me in the JIRA, I'll change this to match dnf. I'll also add a new flag `--disableexcludes` to exclude those from config file as this looks useful keep around.

sidcha replied:
This is now fixed.

ppadmavilasom commented on client/client.c:
```diff
@@ -77,6 +77,16 @@ TDNFPkgsToExclude(
 
+    if (pTdnf->pConf->ppszExcludes)
+    {
+        printf("Warning: The following packages are excluded from tdnf.conf:\n");
```
you might have to check if `quiet` flag is on before printing this

sidcha replied:
Will fix this, thanks. Should we have a method for logging in TDNF (that takes pTdnf)? that can handle these nits instead of this Ad-Hoc prints? I can introduce one if the idea sounds okay.

ppadmavilasom replied:
sounds good.

sidcha replied:
Checked for quite in this instance. Will raise a separate PR for logging.

ppadmavilasom commented on common/strings.c:
```diff
@@ -88,6 +88,57 @@ TDNFSafeAllocateString(
     goto cleanup;
 }
 
+uint32_t
```
if split into a count and collect, can avoid realloc

sidcha replied:
Here is my justification for using `realloc()` in this instance. I'm open for suggestions/corrections:

- The number of realloc() calls are logarithmic to the target size and we start off with a large chunk (16 in this case) and double each time.
- `strtok()` (or other tokenizing methods `strsep()` etc) cobblers the input string so I'd have to make two copies (`strdup()`-ed) to achieve this effect. This makes the code look ugly :).
- The talingling resize realloc() does no copy or dealloc most of the time.
- This progressive growth pattern is easy understand and looks good.

ppadmavilasom replied:
you dont have to use strtok. you have some separator - space in this case. just count space instances using an advancing ptr and strchr. i think its much cleaner to know what you need to do and then allocate precisely for that.
again, just my opinion.

sidcha replied:
Switched to count-and-alloc

oliverkurth commented on common/strings.c:
```diff
@@ -88,6 +88,91 @@ TDNFSafeAllocateString(
     goto cleanup;
 }
 
+size_t
```
Common style in tdnf code is to return dwError, and return the value in an argument.

sidcha replied:
This method has no errors. An error here is application/usage dependant. For instance someone can pass a string and  check if it does not have any of the give separator chars by check `count == 1`.

The mandate of using `BAIL_ON_TDNF_ERROR` is causing methods that don't need to return anything also return a `uint32_t` just so the caller can also call `BAIL_ON_TDNF_ERROR`. It's fast becoming an anti-pattern IMO. But do let me know if this is a "must be followed".

ppadmavilasom replied:
https://github.com/vmware/tdnf/blob/dev/docs/coding-guidelines.md
Credit to @snambakam

sidcha replied:
Yes, I've seen this. Specifically:

> If a function has the possibility to fail, it must return a 32 bit error code

oliverkurth replied:
I don't really like his, because it's different to the behaviors of similar functions and therefore maybe confusing. Caller may not know that this function cannot fail, and maybe this will change some time in the future.

I won't insist on the change.

sidcha replied:
Changed.

A caller _must_ know know a function's contract before invoking it.

oliverkurth commented on common/strings.c:
```diff
@@ -88,6 +88,91 @@ TDNFSafeAllocateString(
+    if (IsNullOrEmptyString(pszSep))
+    {
+        nCount = 1;
+        BAIL_ON_TDNF_ERROR(1);
```
Here we should use a macros as defined in `include/tdnferror.h `.

Like:
```
dwError = ERROR_TDNF_INVALID_PARAMETER
BAIL_ON_TDNF_ERROR(dwError);
```

tapakund replied:
Is this resolved? Current changes still shows we passing hardcoded value.

sidcha replied:
Actually `BAIL_ON_TDNF_ERROR` here is misleading. It must just be `goto exit;` as this method does not return have an erroring scenario (see thread above).

```c
if (IsNullOrEmptyString(pszBuf))
{
    goto exit;
}
if (IsNullOrEmptyString(pszSep))
{
    nCount = 1;
    goto exit;
}
```
I am not sure if tDNF allows bare word goto to be placed instead of a `BAIL_ON_XXX()`. If it does, I'll go ahead with that.

oliverkurth commented on common/strings.c:
```diff
@@ -88,6 +88,91 @@ TDNFSafeAllocateString(
+    }
+
+error:
+    return nCount;
```
return dwError here.

sidcha committed:
Hoist call to TDNFPkgsToExclude() in TDNFGoal()

TDNFGoal() calls TDNFPkgsToExclude() in two places. One of these calls
is from within a loop that calls TDNFAddGoal(). Hoisting this call to
TDNFPkgsToExclude() and passing the ouptput to TDNFAddGoal() prevents
needless calls at each loop iteration.

Signed-off-by: Siddharth Chandrasekaran <csiddharth@vmware.com>

sidcha committed:
Add CLI option --disableexcludes to skip config key excludepkgs

Signed-off-by: Siddharth Chandrasekaran <csiddharth@vmware.com>

sidcha commented:
> Is this option required? Can the user remove it from conf file instead of having this option?

DNF has it so I added it for sake of completeness. And yes it might be useful when you want to install excluded packages in a more controlled manner (without having to modify tdnf.conf). I can think of kernel and other core packages that might fall under this category.

ppadmavilasom commented on common/memory.c:
```diff
@@ -62,6 +62,53 @@ TDNFAllocateMemory(
 }
 
+uint32_t
+TDNFReAllocateMemory(
```
unused after your recent refactors.

sidcha replied:
Removed. Thanks.

ppadmavilasom commented on common/strings.c:
```diff
@@ -88,6 +88,91 @@ TDNFSafeAllocateString(
+    }
+
+    pszTemp = pszBuf;
+    while (*pszTemp)
```
offset = 0;
while((next = strstr(str + offset, sep)))
{
      int len = next - (str + offset);
      //allocate array[arrayindex] element with the len above
      memcpy(array[arrayindex], str + offset, len);
      offset = next - str + strlen(sep);
      ++arrayindex;
}
//do an extra allocatestring with source as str+offset

sidcha replied:
The code posted is copying sections too, but I'm doing this in `TDNFSplitStringToArray()`. This method only gives me a leading count on number of sections that is used to allocate memory upfront (to address your review comment earlier).

BTW, the above code does not handle whitespace/multiple-delimiters in input string correctly while mine does.

sshedi commented:
Is it possible to wait for a while? I want to understand this better.

keerthanakalyan commented on tools/cli/lib/help.c:
```diff
@@ -57,6 +57,7 @@ TDNFCliShowHelp(
     printf("           [--reboot-required]\n");
     printf("           [--skipsignature]\n");
     printf("           [--skipdigest]\n");
+    printf("           [--disableexcludes]");
```
Needs a newline after --disableexcludes.

sidcha replied:
Good catch. Thanks.

sidcha commented:
@sshedi, have you had a chance to look into this?

tapakund commented on common/configreader.c:
```diff
@@ -593,6 +593,47 @@ TDNFReadKeyValueInt(
+    char** ppszValList = NULL;
+    PKEYVALUE pKeyValues = NULL;
+
+    if(!pSection || !pszKeyName || !pppszValueList)
```
maintain space between if (

sidcha replied:
Done.

tapakund commented on common/configreader.c:
```diff
@@ -593,6 +593,47 @@ TDNFReadKeyValueInt(
+    pKeyValues = pSection->pKeyValues;
+    for(; pKeyValues; pKeyValues = pKeyValues->pNext)
+    {
+        if(strcmp(pszKeyName, pKeyValues->pszKey) == 0)
```
same here.

sidcha replied:
Done.

tapakund commented on common/configreader.c:
```diff
@@ -593,6 +593,47 @@ TDNFReadKeyValueInt(
+    }
+
+    pKeyValues = pSection->pKeyValues;
+    for(; pKeyValues; pKeyValues = pKeyValues->pNext)
```
here

sidcha replied:
Done.

tapakund commented on common/strings.c:
```diff
@@ -88,6 +88,91 @@ TDNFSafeAllocateString(
+    size_t nCount = 0;
+    size_t nLength = 0;
+
+    nCount = TDNFStringSepCount(pszBuf, pszSep);
```
Better to NULL check the parameters before passing it on to a function.

tapakund commented on common/strings.c:
```diff
@@ -88,6 +88,91 @@ TDNFSafeAllocateString(
+    pszTok = strtok_r(pszBuf, pszSep, &pszState);
+    while (nLength < nCount && pszTok != NULL)
+    {
+        dwError = TDNFAllocateString(pszTok, &ppszToks[nLength]);
```
i think renaming nLength to nIndex makes more sense while going through the code. its just a suggestion, you can take a call.

sidcha replied:
Done.

sshedi reviewed (approved):
LGTM

keerthanakalyan reviewed (approved):
LGTM

oliverkurth commented:
LGTM.

sidcha committed:
Make TDNFStringSepCount() follow TDNF's dwError pattern

Signed-off-by: Siddharth Chandrasekaran <csiddharth@vmware.com>
